### PR TITLE
Password strength indicator

### PR DESCRIPTION
**Pull request by simeonbizhub**
[Trello](https://trello.com/c/aC7BzeWN)

These changes include an implementation of `zxcvbn` (JavaScript library) for:
- `changePassword.jsp` (when 'must change password' = true)
- `resetPassword.jsp` (password reset)
- `selfRegistration.xhtml` (self registration)

**Changes**
- Implemented strength indicator for `resetPassword.jsp` to `changePassword.jsp`
- Added custom view (`selfRegistration.xhtml`) for `admin.SelfRegistration` & implemented same here
- Added `strengthIndicator.css` for common strength bar stylings

### CODE REVIEW

**Comment by benpetito** [comment]
Hey @simeonbizhub , there are conflicts on this branch sorry, are you able to resolve those?

File: skyve-war/src/main/webapp/external/selfRegister.xhtml [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+		<p:importEnum type="org.skyve.web.WebAction" var="WebAction" allSuffix="ALL_ENUM_VALUES" />  
+		<ui:composition template="#{'/WEB-INF/pages/templates/'.concat(skyvePublic.templateName).concat('/view.xhtml')}">
+			<ui:param name="bean" value="#{skyvePublic}" />
+			<ui:param name="showMenu" value="false" />
+			<ui:param name="resetMenu" value="false" />
+			<ui:param name="showSwitchMode" value="#{'phone' ne skyve.userAgentType and skyve.canSwitchMode}" />
+			<ui:param name="showLogout" value="false" />
+			<ui:define name="head">
+				<!-- Password strength estimator -->
+				<script src="https://cdnjs.cloudflare.com/ajax/libs/zxcvbn/4.2.0/zxcvbn.js"></script>
```
Comment (benpetito) [inline skyve-war/src/main/webapp/external/selfRegister.xhtml]: Can you download this file and add it into Skyve where we have the other javascript files? That allows this to work offline and protects us from CDN poisoning attacks.

File: skyve-war/src/main/webapp/pages/resetPassword.jsp [hunk 2]
```diff
@@ -143,6 +147,67 @@
...
+				    3: "75%",
+				    4: "100%"
+				}
+			var progressBarColour = {
+				    0: "#D73F40",
+				    1: "#DC6551",
+				    2: "#F2B84F",
+				    3: "#BDE952",
+				    4: "#3ba62f"
+				}
```
Comment (benpetito) [inline skyve-war/src/main/webapp/pages/resetPassword.jsp]: I would move this into a common JS file so it's defined in one place.

File: skyve-war/src/main/webapp/external/selfRegister.xhtml [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+						widgetId="actions" />
+					
+					<!-- Registration complete -->
+					<s:view module="#{skyvePublic.bizModuleParameter}"
+						document="#{skyvePublic.bizDocumentParameter}"
+						managedBean="skyvePublic"
+						update="@form"
+						widgetId="postRegister" />
+								
+					<div id="outerPanelId" widgetVar="doRegister" style="border: none;">
```
Comment (benpetito) [inline skyve-war/src/main/webapp/external/selfRegister.xhtml]: `widgetVar` is not a valid HTML attribute of `div`, does this do anything here?

File: skyve-war/src/main/java/modules/admin/router.xml [hunk 4]
```diff
@@ -3,7 +3,7 @@
 			xsi:schemaLocation="http://www.skyve.org/xml/router ../../schemas/router.xsd"
 			xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
 	<uxui name="external">
-		<route outcome="/external/public.xhtml">
+		<route outcome="/external/selfRegister.xhtml">
```
Comment (benpetito) [inline skyve-war/src/main/java/modules/admin/router.xml:6]: I can't get this to work, it never routes for me from the login page. It just flashes and goes nowhere.

**Review by benpetito — CHANGES_REQUESTED** [review]

**Comment by benpetito** [comment]
[Review 'Password strength indicator' PR](https://trello.com/c/JZdv6JXA/75-review-password-strength-indicator-pr)

File: skyve-war/src/main/webapp/external/selfRegister.xhtml [hunk 5]
```diff
@@ -26,8 +26,9 @@
 			<ui:param name="showLogout" value="false" />
 			<ui:define name="head">
 				<!-- Password strength estimator -->
-				<script src="https://cdnjs.cloudflare.com/ajax/libs/zxcvbn/4.2.0/zxcvbn.js"></script>
-                <link rel="stylesheet" type="text/css" href="pages/css/strengthIndicator.css" />
+				<script type="text/javascript" src="strength-indicator/zxcvbn.js"></script>
+				<script type="text/javascript" src="strength-indicator/util.js"></script>
+				<link rel="stylesheet" type="text/css" href="pages/css/strength-indicator.css"></link>
```
Comment (benpetito) [inline skyve-war/src/main/webapp/external/selfRegister.xhtml]: Can this file be moved into the skyve-web jar and in the same folder as the zxcvbn and util js? Then it doesn't need to be part of the admin module.

**Review by benpetito — CHANGES_REQUESTED** [review]
Hey @simeonbizhub , one last tiny change sorry!

**Comment by benpetito** [comment]
Hi @simeonbizhub ,

could you please:

* minify the CSS and JS you added
* use the HIBP hash to track the password attempt instead of keeping the full hashed new password in the session

**Comment by simeonbizhub** [comment]
> Hi @simeonbizhub ,
> 
> could you please:
> 
> * minify the CSS and JS you added
> * use the HIBP hash to track the password attempt instead of keeping the full hashed new password in the session

Thanks @benpetito, done.

File: skyve-war/src/main/webapp/external/selfRegister.xhtml [hunk 6]
```diff
@@ -75,6 +72,10 @@
 									childStyle="padding: 0px;"
 									childStyleClass="ui-g-12" />
 									
+								<!-- Password strength estimator -->
+								<script type="text/javascript" src="zxcvbn/zxcvbn-4.4.2-min.js"></script>
+								<link type="text/css" rel="stylesheet" href="zxcvbn/strength-bar-min.css"/>
+
```
Comment (benpetito) [inline skyve-war/src/main/webapp/external/selfRegister.xhtml]: Why are these imports with the view markup? I would put these with the other scripts at the bottom of the page? ~ Line 120

**Review by benpetito — CHANGES_REQUESTED** [review]
One last question, thanks Simeon.

File: skyve-war/src/main/webapp/external/selfRegister.xhtml (hunk 6 above)
Comment (simeonbizhub) [reply skyve-war/src/main/webapp/external/selfRegister.xhtml]: No reason - just moved :)

**Pull request merged**